refactoring SpotClient to match original code base better

mlake opened the pull request:
correctly identifying some public spot api endpoints

added two unit tests for spot to get started

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes # <!-- Issue # here -->

## Summary
<!--
Thanks for creating this pull request.

Please make sure that the pull request is limited to one type (enhancement, bug, etc.) and keep it as small as possible. You can open multiple PRs instead of opening a huge one.
-->

<!-- Add a brief description of the pr -->

<!-- You can also choose to add a list of changes and if they have been completed or not by using the markdown to-do list syntax
- [ ] Not Completed
- [x] Completed
-->

## Additional Information
<!-- Any additional information like breaking changes, dependencies added, screenshots, comparisons between new and old behavior, etc. -->

##  Checks
<!-- Make sure your pr passes the CI checks and do check the following fields as needed - -->
- [ ] Updated related documentation
- [ ] Updated/added related tests
- [ ] I have tested my changes locally

mlake committed:
refactoring SpotClient to match original code base better

correctly identifying some public spot api endpoints

added two unit tests for spot to get started

mlake committed:
unit tests added for all public SPOT api endpoints

untested: non-public API endpoints

tiagosiebler commented on src/spot-client.ts:
```diff
@@ -1,157 +1,153 @@
-  getBalances() {
-    return this.getPrivate('/spot/v1/account');
-  }
+export class SpotClient {
```
I'm not keen in restoring RequestWrapper (and removing the extended BaseRestClient). It was an intentional decision to remove it and instead extend a shared BaseRestClient.

tiagosiebler commented on src/spot-client.ts:
```diff
@@ -1,157 +1,153 @@
+import {AxiosRequestConfig} from 'axios';
+import {KlineInterval, APIResponse, SymbolInfo} from './types/shared';
+import {NewSpotOrder, OrderSide, OrderTypeSpot, SpotOrderQueryById} from './types/spot';
+import {GenericAPIResponse, getRestBaseUrl, RestClientOptions} from './util/requestUtils';
```
Lots of style changes here, please try to preserve the existing styling (2-spaced indent, 1 space around the braces)

tiagosiebler commented on package.json:
```diff
@@ -39,7 +39,8 @@
     "webpack-bundle-analyzer": "^4.1.0",
-    "webpack-cli": "^4.2.0"
+    "webpack-cli": "^4.2.0",
+    "axios-debug-log": "^0.8.4"
```
Don't forget to remove this when done

mlake replied:
any particular reason? it's a dev dependency that doesn't show up in the final code base and only shows up in dev/unit testing

tiagosiebler replied:
Probably overkill on my end but I'm trying to keep dependencies as trimmed as possible. If it's something more complicated yet useful/essential, then I'm open to evaluate it, but for smaller features that could also be done in a few lines of code I'd rather just make an implementation of it instead of introducing another dependency, even if dev.

Here's an example equivalent you could wrap in a function and call when you want logging for requests/responses:
```
axios.interceptors.request.use(request => {
  console.log('Starting Request', JSON.stringify(request, null, 2));
  return request;
});

axios.interceptors.response.use(response => {
  console.log('Response:', JSON.stringify(response, null, 2));
  return response;
});
```

tiagosiebler commented on src/util/requestUtils.ts:
```diff
@@ -51,11 +51,17 @@ export function getRestBaseUrl(useLivenet: boolean, restInverseOptions: RestClie
-  if (endpoint.startsWith('public/linear')) {
-    return true;
+
+  const publicPrefixes = [
+    'v2/public',
+    'public/linear',
+    'spot/quote/v1',
+    'spot/v1/symbols',
+    'spot/v1/time'
+  ]
+
+  for (const prefix of publicPrefixes) {
+    if (endpoint.startsWith(prefix)) return true
```
Looks better this way 👍 don't forget the semi-colons

tiagosiebler replied:
Still missing semi colons

tiagosiebler commented on test/spot/public.test.ts:
```diff
@@ -0,0 +1,57 @@
+
+        it('getOrderBook()', async () => {
+            expect(await api.getOrderBook(symbol)).toMatchObject(successResponseObjectSpot());
+        });
```
Looks great! key and secret aren't being used yet but I guess it's just public endpoints at first anyway. Please just fix:
- remove the axios debug log dependency & import
- update the indent (2 spaces)
- add semi colons where missing

mlake replied:
updated the indentation. You should probably use prettier + husky for code style consistency

tiagosiebler replied:
Definitely - haven't gotten to setting it up yet

tiagosiebler commented on test/spot/public.test.ts:
```diff
@@ -0,0 +1,57 @@
+import {SpotClient} from "../../src/spot-client";
+import {successResponseListSpot, successResponseObjectSpot} from "../response.util";
```
```suggestion
import { SpotClient } from "../../src/spot-client";
import { successResponseListSpot, successResponseObjectSpot } from "../response.util";
```

nit: padding around braces

mlake committed:
added back the currently unused BaseRestClient.ts file

Spot: Fixed invalid signature for authenticated calls
   there's a subtle difference for recv_window between the API endpoints

   removed "undefined" params that were being serialized
   SPOT endpoint requires different time server

Unit tests:
  added all spot public calls
  added some authenticated (read-only) tests for Inverse and Spot can supply BYBIT_KEY and BYBIT_SECRET as env variables

mlake committed:
adding missing startTime and endTime parameters for getMyTrades

tiagosiebler commented on src/spot-client.ts:
```diff
@@ -18,62 +17,62 @@ export class SpotClient extends BaseRestClient {
-  async getServerTime(urlKeyOverride?: string): Promise<number> {
-    const result = await this.get('/spot/v1/time');
-    return result.serverTime;
+  getServerTime(): GenericAPIResponse {
+    return this.requestWrapper.get('spot/v1/time');
```
This should return a promise resolving a number,

tiagosiebler commented on src/spot-client.ts:
```diff
@@ -1,11 +1,10 @@
 import RequestWrapper from './util/requestWrapper';
 
-export class SpotClient extends BaseRestClient {
+export class SpotClient {
```
```suggestion
export class SpotClient extends BaseRestClient {
```

I want to keep extending base rest client, so it can enforce implementation of getServerTime() resolving to a number, as well as being able to make requests with this.get/post/delete/etc()

tiagosiebler commented on src/spot-client.ts:
```diff
@@ -18,62 +17,62 @@ export class SpotClient extends BaseRestClient {
-    return this.get('/spot/v1/symbols');
+
+  getSymbols(): Promise<APIResponse<SymbolInfo[]>> {
+    return this.requestWrapper.get('spot/v1/symbols');
```
```suggestion
    return this.get('spot/v1/symbols');
```

Request wrapper should be deprecated, please revert this to `this.get('spot/v1....`

tiagosiebler commented on src/util/BaseRestClient.ts:
```diff
@@ -13,11 +19,11 @@ export default abstract class BaseRestClient {
-    baseUrl: string,
-    options: RestClientOptions = {},
-    requestOptions: AxiosRequestConfig = {}
+      key: string | undefined,
+      secret: string | undefined,
+      baseUrl: string,
+      options: RestClientOptions = {},
+      requestOptions: AxiosRequestConfig = {}
```
Indent should be 2 spaces

tiagosiebler commented on src/util/requestWrapper.ts:
```diff
@@ -77,6 +80,16 @@ export default class RequestUtil {
+      // Optional, set to 5000 by default. Increase if timestamp/recv_window errors are seen.
+      if (this.options.recv_window) {
+
+        if (endpoint.includes("spot/")) { // lucky us. only the spot API uses a different param for this
+          params = { ...params, recvWindow: this.options.recv_window };
+        } else {
+          params = { ...params, recv_window: this.options.recv_window }
+        }
+      }
```
Yikes, if only they were consistent... good catch

tiagosiebler replied:
I just realised that the my bybit client classes aren't using the BaseRestClient yet, which explains why you've been using the requestWrapper too.

All my connectors (bybit, ftx, binance) follow similar design patterns. Binance is my most recent and probably more improved iteration of these patterns, where I decided to move all request logic to BaseRestClient, which is extended by the REST client. This makes the dependency on the getTime endpoint less circular and solved a bunch of other problems. Eventually I wanted to migrate to that design with this bybit and also the ftx connector.

You can see the implementation here:
https://github.com/tiagosiebler/binance/blob/master/src/main-client.ts#L90-L107

Could you migrate the request logic into the BaseRestClient in this PR? It doesn't need to be a full migration on the inverse/linear/futures clients yet, that can come separately since that's probably no small feat.

I would start with migrating what's currently in the bybit-api requestWrapper into the BaseRequestClient, then apply your improvements to it to also support signed requests to the spot apis, then update the spot client to use the BaseRequestClient similarly to how the REST clients in my binance connector do it.

Then when it comes to later migrating the other bybit-api REST clients to extend the BaseRequestClient it'll be less work, since it has an extension of the logic already in the requestWrapper. Does this make sense? Feel free to ping me on telegram if you have questions/thoughts/concerns.

tiagosiebler commented on test/inverse/public.test.ts:
```diff
@@ -5,19 +5,33 @@ describe('Public Inverse REST API Endpoints', () => {
   const from = Number(timestampOneHourAgo.toFixed(0));
 
+
+  describe('Inverse Authenticated Read-Only Endpoints', () => {
+
+    it('getTrades()', async () => {
+      let trades = await apiAuth.getTrades({ symbol });
+      expect(trades).toMatchObject(successResponseObject());
+    });
+
+  });
```
I'd prefer a new test file dedicated to private tests (`test/inverse/private.test.ts`)

tiagosiebler commented:
Thanks again for the PR - closing this in favour of #150 , which includes some of your fixes.
